CoinDesk trend indicator rebalance bot ($cdETI)

edkim opened the pull request:
Sets up Tenderly action that periodically
1)  Fetches the CoinDesk API for ETH trend score and 
2) Compares the score with the current positioning of cdETI (bearish, neutral, or bullish) 
3) Executes a Uniswap trade if the current positioning (e.g. 100% WETH) doesn't align with the trend score (e.g. -0.5).

Intended to be deployed to Tenderly Web3Actions using `tenderly-cli`

janndriessen commented on .env.default:
```diff
@@ -9,3 +9,5 @@ INDEX_0X_API=
 MAINNET_RPC_URL=
 TENDERLY_FORK_URL=
+TREND_INDEX_REBALANCER_PRIVATE_KEY=
+COINDESK_API=
```
There is a couple of `no new line at end of file`. Usually, this is just a setting in the editor to fix and it will be automatically done.

edkim replied:
Good tip. Changed it in my VSCode.

janndriessen commented on actions/lib/trendIndexRebalance.ts:
```diff
@@ -0,0 +1,213 @@
+      uniV3AdapterWethToUsdc.fees,
+      uniV3AdapterWethToUsdc.fixIn
+    );
+	}
```
minor: could return values right away instead of storing in var

janndriessen commented on test/trendIndexRebalance.test.ts:
```diff
@@ -0,0 +1,167 @@
+    });
+    return ethers.provider.getSigner(address);
+  }
+});
```
minor: might be a good thing for utils if it's reused across tests?

edkim commented on actions/lib/trendIndexRebalance.ts:
```diff
@@ -0,0 +1,205 @@
+    sendToken,
+    sendQuantity,
+    receiveToken,
+    0,
```
My intention is to always use Flashbots RPC. Is it worth fetching price and enforcing 0.5% slippage here? @ckoopmann @snake-poison

edkim commented on test/trendIndexRebalance.test.ts:
```diff
@@ -0,0 +1,175 @@
+    weth = IWETH__factory.connect(wethAddress, signer);
+  });
+
+  // Test if mock server returns expected values
```
This feels like too much code for what I'm trying to do. I'd like to come back to this when my brain is fresh.

edkim replied:
I ended up hoisting the Coingecko API call out of the `tryTrendIndexRebalance` function and removing the mock api tests.

ckoopmann replied:
That means the api fetching / trend score calculation is not tested at all anymore though, right ?

edkim replied:
Correct about no longer testing the API fetching. Is there a reason we'd want to test the fetching through an Axios mock? 

If the live CoinDesk API is down or they change the response format, mock tests wouldn't catch it.

ckoopmann replied:
The part of the code you'd be testing would be the handing of the response not the coindesk api. I.e. that your code correctly calculates the trend score from the api response.

I'm not sure how much logic there is for that and how complex it is though, so I'd leave it up to you if you want to that part of your code

ckoopmann reviewed (changes requested):
Good work. 
Went a bit more strict / nit picky then normal here, because you asked for it😅

Also ofc, please fix CI before consider merging and probably merge in cbbeth branch first and then update this one.

janndriessen commented on .github/workflows/test.yml:
```diff
@@ -9,6 +9,7 @@ on:
 env:
   ARB_KEEPER_ADDRESS: ${{ secrets.ARB_KEEPER_ADDRESS }}
   ARB_KEEPER_PRIVATE_KEY: ${{ secrets.ARB_KEEPER_PRIVATE_KEY }}
+  COINDESK_API: ${{ secrets.COINDESK_API }}
```
Just to make sure, you don't need `TREND_INDEX_REBALANCER_PRIVATE_KEY` or `COINDESK_API` here?

edkim replied:
Added, thanks!

janndriessen commented on test/actions/cdeti.ts:
```diff
@@ -0,0 +1,36 @@
+import { rebalance } from "../../actions/cdeti";
+
+dotenv.config();
+
```
I think the test run time may require it. It's the recommended way from the docs: https://docs.tenderly.co/web3-actions/references/local-development-and-testing. Feel free to try with just using jest tho.

edkim replied:
Tenderly docs recommend we run the action from within our Jest tests, but I don't think that adds much value since we can just call tryArb or tryRebalance [sic] directly. 

It does provide an easy way to trigger the action from CLI, which came in handy while developing.

janndriessen commented on actions/lib/trendIndexRebalance.ts:
```diff
@@ -0,0 +1,215 @@
+        shouldRebalance = true;
+        isWethBuy = false;
+        wethComponentUnits = await trendToken.getTotalComponentRealUnits(wethAddress);
+        sendQuantity = wethComponentUnits.div(2);
```
Does it mean selling half of the WETH? To be 50% WETH 50% USDC in the neutral strategy.

janndriessen replied:
As suggested above by Christian it's probably good to either add a few comments or break up the function into multiple smaller _config_/_strategy_ functions

edkim commented on tenderly.yaml:
```diff
@@ -58,6 +58,15 @@ actions:
+          block:
+            network:
+              - 1
+            blocks: 1200
```
The CoinDesk API only gets updated once per day, so this should be fine. 

Separately, Allan requested that any rebalance transactions be done at a random time within 24 hours of the trend score being updated. I haven't thought of a good way to do that yet. Any ideas @FlattestWhite?

edkim commented on package.json:
```diff
@@ -9,6 +9,8 @@
     "bot-deploy": "tenderly actions build && tenderly actions deploy",
     "bot-test": "ts-node test/actions/index.ts",
+    "cdeti-rebalance-action": "ts-node test/actions/cdeti.ts",
+    "cdeti-rebalance-test": "hardhat test test/trendIndexRebalance.test.ts",
```
Removed!

edkim commented on test/actions/cdeti.ts:
```diff
@@ -0,0 +1,36 @@
+  blockEvent.network = "1";
+  blockEvent.blockHash =
+    "0xccc2b2ba35e4965dcfcfcafd4814a42b2f542aa405e85818da4eb5aa554c4324";
+  blockEvent.blockNumber = 18220964;
```
I deleted these blockEvent lines since I don't know if they're needed. Let's see if Tenderly complains.

edkim commented on test/trendIndexRebalance.test.ts:
```diff
@@ -0,0 +1,175 @@
+      });
+    });
+
+    describe("and trend token is 100% usdc", function () {
```
I've tried to make it more clear that before certain tests, we run `tryTrendIndexRebalance()` with a hardcoded trend score, to trade into the token composition we want to test against.

Perhaps a better way is to avoid using `tryTrendIndexRebalance()` in our tests, and instead use helper functions to achieve the fork state we want.

edkim commented on test/trendIndexRebalance.test.ts:
```diff
@@ -0,0 +1,175 @@
+        console.log('usdc balance after', usdcBalanceAfter.toString());
+        console.log('weth balance after', wethBalanceAfter.toString());
+        
+        expect(wethBalanceAfter).to.be.gt(wethBalanceBefore);
```
I think it's enough. We could add a price oracle check but don't think it's worth it.

ckoopmann replied:
Okay. Up to you.

edkim commented on actions/lib/utils.ts:
```diff
@@ -1,16 +1,31 @@
-    return new Wallet(arbKeeperPrivateKey, provider);
+export async function getSigner(context: Context, secret: string) {
+    const privateKey = await context.secrets.get(secret);
+    const provider = new providers.JsonRpcProvider(rpcUrl, 1);
```
Note that this replaces an environment secret for JSON-RPC URL with a hardcoded Flashbots RPC URL. This also means that the icreth and iceth web3actions will also now use Flashbots.

ckoopmann replied:
I think it is better to keep it in the secret and just set the secret value to the flashbots rpc. That way we are able to change the rpc if we want to by just updating the secret. (if we keep it like this we need to adjust the whole code and redeploy the actions).

ckoopmann replied:
Also the flashbots rpc is pretty prone to rate limiting which I had a lot of issues with on my own arb bots.
We probably want to configure it such that we only use flashbots for the final transaction and not all of the simulation calls before

ckoopmann replied:
It is also slower than alchemy and does not give us the logging. Another reason why we would want to use it only where we need to.

edkim replied:
Ah good thing you know from experience. I'll split it back out as a separate Secret. I think it's good to be able to change the RPC without re-deploying...but I don't love that there's no record of what the secret has been changed to/from, which could make it harder to debug later.

ckoopmann replied:
You could include the rpc url in the logs for debugging.

ckoopmann commented on actions/lib/trendIndexRebalance.ts:
```diff
@@ -0,0 +1,215 @@
+
+  const trendScore = await fetchTrendScore(apiUrl);
+
+  switch (trendScore) {
```
Yes, definetely more readable this way👏

ckoopmann reviewed (changes requested):
Mostly just added additional comments / answers to previously raised nitpicks. All of those are optional though, will leave it up to you in the end. 

One point I would definetely change though is not to use the Flashbots RPC everywhere. This  url is much slower when running the different simulations etc. and rate limits pretty aggressively so it should only be used where it is actually needed. (which is when submitting transactions that are subject to potential frontrunning). I had my personal arb bots repeatedly hang / fail to execute due to this.

Also we should fix flaky test first before merging. Lmk if you need my support there. (it is one of the tests that I added that is flaking after all😅)

Overall great work despite my nit picking here and there🙏

edkim commented:
@ckoopmann I've updated the PR with
- Using Flashbots RPC only for executing rebalance Tx
- Deleted separate `test/actions/cdeti.ts` test file, and combined the runtime test with the other hardhat tests, as you suggested.
- Added Discord notifications, which uses Tenderly storage so that we can change alert subscribers easily.

ckoopmann reviewed (approved):
Great Work! Thanks for the patience and diligence!
LGTM + LFG !!!

pblivin0x commented on actions/lib/trendIndexRebalance.ts:
```diff
@@ -0,0 +1,190 @@
+  );
+}
+
+async function getUniV3SwapData(
```
potential to add slippage protection here?
